Restore thread interrupt flag after an InterruptedException

danielmitterdorfer opened the pull request:
This commit replaces all occurrences of Thread.interrupted() with
Thread.currentThread().interrupt(). While the former checks and clears the current
thread's interrupt flag the latter sets it, which is actually intended.

Closes #14798

danielmitterdorfer commented:
I've looked in the JDK source of `AbstractExecutorService`, `FutureTask` and similar classes. The standard implementation `ThreadPoolExecutor` ensures in the method `runWorker()` that the interrupt flag is cleared unless the pool is being shut down in which case the thread is actively interrupted.

The best reference on how to implement interruption policies in Java can be found in Java Concurrency in Practice, which basically advises that you should restore the flag or rethrow `InterruptedException` whenever you do not know what code is higher up the call stack (i.e. every time you do not extend `java.lang.Thread`).

@bleskes I am going to comment on each usage where the flag is restored soon.

danielmitterdorfer commented on core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java:
```diff
@@ -348,7 +348,7 @@ public void onFailure(Throwable e) {
                 success = true;
             } catch (InterruptedException e) {
-                Thread.interrupted();
+                Thread.currentThread().interrupt();
```
As this class is used on the client, which we do not control at all, this should be treated as library code. As library we do not know and should also not assume in which environment we are running and how the client wants to respond to interruption. Hence, we should definitely call `Thread.currentThread().interrupt()` here. However, we should restore the flag at the end of the catch block in order to reduce potential side-effects in listeners which we also do not control.

bleskes replied:
In this case I would argue that

> rethrow InterruptedException whenever you do not know what code is higher up the call stack

translates to passing the exception to the listener - it's what we do for any other exception.

danielmitterdorfer replied:
This puts the burden on the client (or anybody that implements such a listener), which needs to know that we do not restore the interrupt flag and do it in the listener implementation instead. As a consequence of this decision the client will not be able to cancel a bulk request *unless* it is aware of the idiosyncrasies of our client API.

The problem is that `InterruptedException` is not like any other exception and we're really interfering with the expectations of the client here. To me we're assuming too much about the client when we do not restore the flag here.

danielmitterdorfer replied:
We've discussed that this will go into a separate PR as it directly affects the client and I need to check anyway whether we're handling all cases in `BulkProcessor` consistently (i.e. numberOfConcurrentRequest > 0 vs. numberOfConcurrentRequest == 0)

danielmitterdorfer commented on core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractRunnable.java:
```diff
@@ -36,7 +36,7 @@ public final void run() {
             doRun();
         } catch (InterruptedException ex) {
-            Thread.interrupted();
+            Thread.currentThread().interrupt();
```
As a base class it should stick to standard practice (which is to restore the flag). If subclasses want to do anything differently, they can always catch `InterruptedException` and handle it explicitly.

bleskes replied:
same hare - since we have on onFailure handler, calling is the equivalent of re-throwing the exception, imo.

danielmitterdorfer replied:
To me that's also the same as above but as we're on the server-side here I'm not as reluctant as on the client side to agree though I still think it's broken. The separate catch block for `InterruptedException` indicates that the original developer was aware that we have to treat `InterruptedException` differently. It's just that the handling is not correct.

However, if we don't handle `InterruptedException` differently, I would remove the separate catch block too:

```
    @Override
    public final void run() {
        try {
            doRun();
        } catch (Throwable t) {
            onFailure(t);
        } finally {
            onAfter();
        }
    }
```

danielmitterdorfer commented on core/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -700,7 +700,7 @@ public void processPendingDeletes(Index index, IndexSettings indexSettings, Time
                             logger.debug("{} schedule pending delete retry after {} ms", index, sleepTime);
                         } catch (InterruptedException e) {
-                            Thread.interrupted();
+                            Thread.currentThread().interrupt();
```
This one is a bit more tricky, but there are also other calls that also set the interruption flag already (see `InternalShardLock#acquire()` which is called by `NodeEnvironment#lockAllForIndex()`). After looking what happens in the error handling code for this case, I'd opt for rethrowing `InterruptedException` directly instead of just setting the flag. In conjunction with that this shows that it makes sense that `AbstractRunnable` restores the flag on `InterruptedException` then (this whole code block is enclosed in a subclass of `AbstractRunnable`).

bleskes replied:
Imho, the interruption is dealt with here. We don't need to bother the code higher up with it.

danielmitterdorfer replied:
Agreed. However, as I was trying to explain in my comment, we're dealing with interruption inconsistently in this method. When we're interrupted while we're stuck in `InternalShardLock#acquire()` we throw an exception and will *not* notify the master (i.e. we will not call the second method `TransportService#sendRequest()` from `IndicesService#lockIndexAndAck()`) but if we are interrupted while we are in a timeout we will still notify the master (as we just return). If that's ok, then I'm fine with that. :)

bleskes replied:
OK. I'm good with  fixing this inconsistency by throwing an InterruptedException (and catching it up  just like we do with LockObtainFailedException

danielmitterdorfer commented on core/src/main/java/org/elasticsearch/indices/ttl/IndicesTTLService.java:
```diff
@@ -99,7 +99,7 @@ protected void doStop() {
             this.purgerThread.shutdown();
         } catch (InterruptedException e) {
-            Thread.interrupted();
+            Thread.currentThread().interrupt();
```
This is called in `#doStop()` which is called when the component is about to stop anyway and does no harm.

danielmitterdorfer commented on core/src/main/java/org/elasticsearch/indices/ttl/IndicesTTLService.java:
```diff
@@ -340,7 +340,7 @@ public void await() {
                 condition.await(timeout.millis(), TimeUnit.MILLISECONDS);
             } catch (InterruptedException e) {
-                Thread.interrupted();
+                Thread.currentThread().interrupt();
```
This is used by an internal thread where we control the run loop so the interruption here is also fine. However, I am not sure why we create threads ourselves outside of pools. Btw, I think this thread's implementation could be simplified significantly if we stick to Java's standard methods, i.e. interruption :) (but I would not do this in the scope of this PR).

bleskes replied:
Looking what this does, an interruption is ignored and will only used to cause an extra TTL check or shutdown. I think doing nothing is the right move here.

danielmitterdorfer replied:
Ok, I'll implement an empty catch block and just put a comment there.

danielmitterdorfer commented on test-framework/src/main/java/org/elasticsearch/test/CompositeTestCluster.java:
```diff
@@ -73,7 +73,7 @@ public synchronized void beforeTest(Random random, double transportClientRatio)
                     externalNodes[i] = externalNodes[i].start(client, defaultSettings, NODE_PREFIX + i, cluster.getClusterName(), i);
                 } catch (InterruptedException e) {
-                    Thread.interrupted();
+                    Thread.currentThread().interrupt();
```
Called in `#beforeTest()` so would be ok for me to restore the flag here.

bleskes replied:
I think the test should fail. We don't know what else have happened due to this interrupt nor whether the node is ready.

danielmitterdorfer replied:
Then I think the most sensible thing is to propagate.

danielmitterdorfer commented on test-framework/src/main/java/org/elasticsearch/test/ExternalNode.java:
```diff
@@ -217,7 +217,7 @@ synchronized void stop() {
                     process.waitFor();
                 } catch (InterruptedException e) {
-                    Thread.interrupted();
+                    Thread.currentThread().interrupt();
```
Called in the `#stop()` method which can be interrupted when waiting for an external process to stop. I'd opt for rethrowing InterruptedException were we not implementing `Closeable` and delegating to `#stop()` in our implementation of `#close()`. Even then I'd just rethrow and restore the flag just in `#close()`. Client code does not need to be adapted at all as it is already aware of `InterruptedException` now.

bleskes replied:
+1 to throwing the exception.

danielmitterdorfer replied:
I'll change that accordingly.

bleskes commented:
Thx @danielmitterdorfer for digging into this. Left comments.

danielmitterdorfer commented:
@bleskes I've pushed a revised version and have commented on the parts where we still not agree entirely. :)

danielmitterdorfer commented:
@bleskes I've reverted the changes in `BulkProcessor`, removed the superfluous catch block in `AbstractRunnable` and throw `InterruptedException` in `IndicesService`.

On an unrelated note, I've noticed that `NodeIndexDeletedAction#lockIndexAndAck()` uses a timeout of 30 minutes when calling `#processPendingDeletes()` but the log statement a few lines below says "30 seconds". So one of them is wrong. I assume a timeout of 30 minutes is far too long and it should rather be 30 seconds.

bleskes commented on test-framework/src/main/java/org/elasticsearch/test/ExternalNode.java:
```diff
@@ -233,7 +229,11 @@ synchronized boolean running() {
-        stop();
+        try {
+            stop();
+        } catch (InterruptedException e) {
```
did you look at what will happen if we bubble up this exception?

danielmitterdorfer replied:
`#close()' could either be invoked directly or from within a try-with-resources block and I did not come across either.

Unfortunately, bubbling up is not an option as we implement `Closeable` from the JDK. Hence, I sensed restoring the flag is the most sensible option in this case.

bleskes replied:
+1 

> On 18 Nov 2015, at 16:34, Daniel Mitterdorfer <notifications@github.com> wrote:
> 
> In test-framework/src/main/java/org/elasticsearch/test/ExternalNode.java:
> 
> > @@ -233,7 +229,11 @@ synchronized boolean running() {
> >  
> >      @Override
> >      public void close() {
> > -        stop();
> > +        try {
> > +            stop();
> > +        } catch (InterruptedException e) {
> 
> `#close()' could either be invoked directly or from within a try-with-resources block and I did not come across either.
> 
> Unfortunately, bubbling up is not an option as we implement Closeable from the JDK. Hence, I sensed restoring the flag is the most sensible option in this case.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bleskes commented:
LGTM. Left one minor question.

danielmitterdorfer commented:
@bleskes Thanks for your review efforts! I answered your last comment and will merge the PR soon.
